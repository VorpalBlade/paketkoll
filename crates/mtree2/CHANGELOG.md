
# 0.4.1

## Added

 - Added more documentation

## Changed

 - Change some potential panics to errors.

## Removed

 - Removed test that didn't test anyting useful.

## New contributors

 - @lucab
