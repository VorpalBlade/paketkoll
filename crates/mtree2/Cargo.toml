[package]
authors = ["Richard Dodd <richard.o.dodd@gmail.com>", "Arvid Norlander"]
categories = ["parsing", "os::unix-apis", "filesystem"]
description = "A crate for iterating through the entries of an mtree record file."
edition = "2021"
license = "Apache-2.0 OR MIT"
name = "mtree2"
readme = "README.md"
repository = "https://github.com/VorpalBlade/paketkoll"
version = "0.6.7"

[dependencies]
bitflags.workspace = true
faster-hex.workspace = true
memchr.workspace = true
smallvec.workspace = true

[lints]
workspace = true
