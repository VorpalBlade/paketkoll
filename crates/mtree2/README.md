# mtree2

This is a fork of [mtree-rs](https://github.com/derekdreery/mtree-rs) fixing some issues and improving performance. Unfortunately the upstream
has been dead apart from one comment, so a fork was neccesary.

The original README is reproduced below:

---

A library for iterating through entries of an mtree.

> *mtree* is a data format used for describing a sequence of files. Their location is record, along with optional extra values like checksums, size, permissions etc.

 For details on the spec see [mtree(5)].

## Examples

```rust
use mtree2::MTree;
use std::time::SystemTime;

// We're going to load data from a string so this example with pass doctest,
// but there's no reason you can't use a file, or any other data source.
let raw_data = "
/set type=file uid=0 gid=0 mode=644
./.BUILDINFO time=1523250074.300237174 size=8602 md5digest=13c0a46c2fb9f18a1a237d4904b6916e \
     sha256digest=db1941d00645bfaab04dd3898ee8b8484874f4880bf03f717adf43a9f30d9b8c
./.PKGINFO time=1523250074.276237110 size=682 md5digest=fdb9ac9040f2e78f3561f27e5b31c815 \
     sha256digest=5d41b48b74d490b7912bdcef6cf7344322c52024c0a06975b64c3ca0b4c452d1
/set mode=755
./usr time=1523250049.905171912 type=dir
./usr/bin time=1523250065.373213293 type=dir
";
let entries = MTree::from_reader(raw_data.as_bytes());
for entry in entries {
    // Normally you'd want to handle any errors
    let entry = entry.unwrap();
    // We can print out a human-readable copy of the entry
    println!("{}", entry);
    // Let's check that if there is a modification time, it's in the past
    if let Some(time) = entry.time() {
        assert!(time < SystemTime::now());
    }
    // We might also want to take a checksum of the file, and compare it to the digests
    // supplied by mtree, but this example doesn't have access to a filesystem.
}
```

[mtree(5)]: https://www.freebsd.org/cgi/man.cgi?mtree(5)
