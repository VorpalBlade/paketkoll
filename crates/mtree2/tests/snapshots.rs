use insta::assert_debug_snapshot;
use mtree2::MTree;
use std::fs::File;
use std::path::PathBuf;

macro_rules! test_snapshot {
    ($name:ident, $path:expr) => {
        #[test]
        fn $name() {
            let path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join($path);
            let mtree = MTree::from_reader_with_cwd(File::open(path).unwrap(), PathBuf::from("/"));
            let entries: Vec<_> = mtree.collect();
            assert_debug_snapshot!(entries);
        }
    };
}

test_snapshot!(test_gedit, "examples/gedit.mtree");
test_snapshot!(test_xterm, "tests/data/xterm.mtree");
test_snapshot!(test_relative_paths, "tests/data/relative_paths.mtree");
test_snapshot!(
    invalid_double_filename,
    "tests/data/invalid_double_filename.mtree"
);
test_snapshot!(
    test_wrapped_lines,
    "tests/data/relative_paths_wrapped.mtree"
);
