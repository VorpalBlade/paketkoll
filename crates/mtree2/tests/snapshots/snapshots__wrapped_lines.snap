---
source: crates/mtree2/tests/snapshots.rs
expression: entries
---
[
    Ok(
        Entry {
            path: "/.",
            params: Params {
                checksum: None,
                device: None,
                contents: None,
                flags: Some(
                    [
                        110,
                        111,
                        110,
                        101,
                    ],
                ),
                gid: Some(
                    1000,
                ),
                gname: None,
                ignore: false,
                inode: None,
                link: None,
                md5: None,
                mode: Some(
                    FileMode {
                        mode: 509,
                    },
                ),
                nlink: Some(
                    5,
                ),
                no_change: false,
                optional: false,
                resident_device: None,
                rmd160: None,
                sha1: None,
                sha256: None,
                sha384: None,
                sha512: None,
                size: Some(
                    4096,
                ),
                time: Some(
                    SystemTime {
                        tv_sec: 1457644509,
                        tv_nsec: 817998120,
                    },
                ),
                file_type: Some(
                    Directory,
                ),
                uid: Some(
                    1000,
                ),
                uname: None,
            },
        },
    ),
    Ok(
        Entry {
            path: "/./.mtree.go.swp",
            params: Params {
                checksum: None,
                device: None,
                contents: None,
                flags: Some(
                    [
                        110,
                        111,
                        110,
                        101,
                    ],
                ),
                gid: Some(
                    1000,
                ),
                gname: None,
                ignore: false,
                inode: None,
                link: None,
                md5: None,
                mode: Some(
                    FileMode {
                        mode: 420,
                    },
                ),
                nlink: Some(
                    1,
                ),
                no_change: false,
                optional: false,
                resident_device: None,
                rmd160: None,
                sha1: None,
                sha256: None,
                sha384: None,
                sha512: Some(
                    [
                        126,
                        253,
                        56,
                        204,
                        149,
                        123,
                        83,
                        227,
                        213,
                        171,
                        210,
                        94,
                        157,
                        249,
                        189,
                        231,
                        21,
                        165,
                        225,
                        141,
                        193,
                        138,
                        191,
                        168,
                        240,
                        192,
                        234,
                        169,
                        188,
                        160,
                        253,
                        163,
                        47,
                        170,
                        212,
                        151,
                        43,
                        212,
                        217,
                        181,
                        85,
                        39,
                        15,
                        163,
                        179,
                        185,
                        46,
                        106,
                        30,
                        151,
                        173,
                        30,
                        59,
                        255,
                        241,
                        79,
                        201,
                        51,
                        40,
                        64,
                        4,
                        247,
                        232,
                        211,
                    ],
                ),
                size: Some(
                    12288,
                ),
                time: Some(
                    SystemTime {
                        tv_sec: 1457644483,
                        tv_nsec: 833957552,
                    },
                ),
                file_type: Some(
                    File,
                ),
                uid: Some(
                    1000,
                ),
                uname: None,
            },
        },
    ),
    Ok(
        Entry {
            path: "/./.mtree_test.go.swp",
            params: Params {
                checksum: None,
                device: None,
                contents: None,
                flags: Some(
                    [
                        110,
                        111,
                        110,
                        101,
                    ],
                ),
                gid: Some(
                    1000,
                ),
                gname: None,
                ignore: false,
                inode: None,
                link: None,
                md5: None,
                mode: Some(
                    FileMode {
                        mode: 420,
                    },
                ),
                nlink: Some(
                    1,
                ),
                no_change: false,
                optional: false,
                resident_device: None,
                rmd160: None,
                sha1: None,
                sha256: None,
                sha384: None,
                sha512: Some(
                    [
                        42,
                        174,
                        177,
                        229,
                        240,
                        58,
                        86,
                        61,
                        232,
                        106,
                        76,
                        31,
                        114,
                        43,
                        222,
                        64,
                        49,
                        43,
                        190,
                        203,
                        228,
                        80,
                        77,
                        103,
                        251,
                        68,
                        18,
                        64,
                        192,
                        73,
                        230,
                        173,
                        17,
                        41,
                        78,
                        248,
                        52,
                        17,
                        74,
                        5,
                        230,
                        32,
                        163,
                        144,
                        238,
                        159,
                        66,
                        193,
                        199,
                        123,
                        67,
                        191,
                        26,
                        79,
                        45,
                        71,
                        241,
                        40,
                        34,
                        97,
                        131,
                        150,
                        138,
                        51,
                    ],
                ),
                size: Some(
                    12288,
                ),
                time: Some(
                    SystemTime {
                        tv_sec: 1457644509,
                        tv_nsec: 813998114,
                    },
                ),
                file_type: Some(
                    File,
                ),
                uid: Some(
                    1000,
                ),
                uname: None,
            },
        },
    ),
    Ok(
        Entry {
            path: "/./mtree.go",
            params: Params {
                checksum: None,
                device: None,
                contents: None,
                flags: Some(
                    [
                        110,
                        111,
                        110,
                        101,
                    ],
                ),
                gid: Some(
                    1000,
                ),
                gname: None,
                ignore: false,
                inode: None,
                link: None,
                md5: None,
                mode: Some(
                    FileMode {
                        mode: 436,
                    },
                ),
                nlink: Some(
                    1,
                ),
                no_change: false,
                optional: false,
                resident_device: None,
                rmd160: None,
                sha1: None,
                sha256: None,
                sha384: None,
                sha512: Some(
                    [
                        45,
                        238,
                        159,
                        80,
                        186,
                        22,
                        124,
                        217,
                        13,
                        79,
                        110,
                        132,
                        148,
                        212,
                        115,
                        92,
                        107,
                        6,
                        76,
                        39,
                        158,
                        164,
                        210,
                        110,
                        210,
                        196,
                        117,
                        82,
                        165,
                        192,
                        184,
                        43,
                        11,
                        220,
                        173,
                        178,
                        20,
                        224,
                        202,
                        145,
                        154,
                        23,
                        104,
                        206,
                        14,
                        50,
                        30,
                        141,
                        27,
                        183,
                        30,
                        172,
                        185,
                        197,
                        180,
                        187,
                        95,
                        212,
                        52,
                        41,
                        103,
                        5,
                        90,
                        4,
                    ],
                ),
                size: Some(
                    1719,
                ),
                time: Some(
                    SystemTime {
                        tv_sec: 1457644447,
                        tv_nsec: 20900074,
                    },
                ),
                file_type: Some(
                    File,
                ),
                uid: Some(
                    1000,
                ),
                uname: None,
            },
        },
    ),
    Ok(
        Entry {
            path: "/./mtree.go~",
            params: Params {
                checksum: None,
                device: None,
                contents: None,
                flags: Some(
                    [
                        110,
                        111,
                        110,
                        101,
                    ],
                ),
                gid: Some(
                    1000,
                ),
                gname: None,
                ignore: false,
                inode: None,
                link: None,
                md5: None,
                mode: Some(
                    FileMode {
                        mode: 436,
                    },
                ),
                nlink: Some(
                    1,
                ),
                no_change: false,
                optional: false,
                resident_device: None,
                rmd160: None,
                sha1: None,
                sha256: None,
                sha384: None,
                sha512: Some(
                    [
                        45,
                        238,
                        159,
                        80,
                        186,
                        22,
                        124,
                        217,
                        13,
                        79,
                        110,
                        132,
                        148,
                        212,
                        115,
                        92,
                        107,
                        6,
                        76,
                        39,
                        158,
                        164,
                        210,
                        110,
                        210,
                        196,
                        117,
                        82,
                        165,
                        192,
                        184,
                        43,
                        11,
                        220,
                        173,
                        178,
                        20,
                        224,
                        202,
                        145,
                        154,
                        23,
                        104,
                        206,
                        14,
                        50,
                        30,
                        141,
                        27,
                        183,
                        30,
                        172,
                        185,
                        197,
                        180,
                        187,
                        95,
                        212,
                        52,
                        41,
                        103,
                        5,
                        90,
                        4,
                    ],
                ),
                size: Some(
                    1719,
                ),
                time: Some(
                    SystemTime {
                        tv_sec: 1457644447,
                        tv_nsec: 12900062,
                    },
                ),
                file_type: Some(
                    File,
                ),
                uid: Some(
                    1000,
                ),
                uname: None,
            },
        },
    ),
    Ok(
        Entry {
            path: "/./mtree_test.go",
            params: Params {
                checksum: None,
                device: None,
                contents: None,
                flags: Some(
                    [
                        110,
                        111,
                        110,
                        101,
                    ],
                ),
                gid: Some(
                    1000,
                ),
                gname: None,
                ignore: false,
                inode: None,
                link: None,
                md5: None,
                mode: Some(
                    FileMode {
                        mode: 436,
                    },
                ),
                nlink: Some(
                    1,
                ),
                no_change: false,
                optional: false,
                resident_device: None,
                rmd160: None,
                sha1: None,
                sha256: None,
                sha384: None,
                sha512: Some(
                    [
                        27,
                        58,
                        178,
                        222,
                        39,
                        89,
                        223,
                        110,
                        125,
                        66,
                        77,
                        5,
                        255,
                        209,
                        212,
                        81,
                        1,
                        253,
                        16,
                        174,
                        134,
                        50,
                        138,
                        186,
                        209,
                        101,
                        126,
                        104,
                        4,
                        74,
                        221,
                        105,
                        216,
                        61,
                        107,
                        50,
                        113,
                        34,
                        80,
                        112,
                        244,
                        244,
                        21,
                        74,
                        124,
                        12,
                        153,
                        80,
                        22,
                        7,
                        127,
                        140,
                        121,
                        105,
                        137,
                        14,
                        6,
                        140,
                        208,
                        165,
                        75,
                        35,
                        244,
                        182,
                    ],
                ),
                size: Some(
                    581,
                ),
                time: Some(
                    SystemTime {
                        tv_sec: 1457644509,
                        tv_nsec: 817998120,
                    },
                ),
                file_type: Some(
                    File,
                ),
                uid: Some(
                    1000,
                ),
                uname: None,
            },
        },
    ),
    Ok(
        Entry {
            path: "/./mtree_test.go~",
            params: Params {
                checksum: None,
                device: None,
                contents: None,
                flags: Some(
                    [
                        110,
                        111,
                        110,
                        101,
                    ],
                ),
                gid: Some(
                    1000,
                ),
                gname: None,
                ignore: false,
                inode: None,
                link: None,
                md5: None,
                mode: Some(
                    FileMode {
                        mode: 436,
                    },
                ),
                nlink: Some(
                    1,
                ),
                no_change: false,
                optional: false,
                resident_device: None,
                rmd160: None,
                sha1: None,
                sha256: None,
                sha384: None,
                sha512: Some(
                    [
                        27,
                        58,
                        178,
                        222,
                        39,
                        89,
                        223,
                        110,
                        125,
                        66,
                        77,
                        5,
                        255,
                        209,
                        212,
                        81,
                        1,
                        253,
                        16,
                        174,
                        134,
                        50,
                        138,
                        186,
                        209,
                        101,
                        126,
                        104,
                        4,
                        74,
                        221,
                        105,
                        216,
                        61,
                        107,
                        50,
                        113,
                        34,
                        80,
                        112,
                        244,
                        244,
                        21,
                        74,
                        124,
                        12,
                        153,
                        80,
                        22,
                        7,
                        127,
                        140,
                        121,
                        105,
                        137,
                        14,
                        6,
                        140,
                        208,
                        165,
                        75,
                        35,
                        244,
                        182,
                    ],
                ),
                size: Some(
                    581,
                ),
                time: Some(
                    SystemTime {
                        tv_sec: 1457644509,
                        tv_nsec: 812998113,
                    },
                ),
                file_type: Some(
                    File,
                ),
                uid: Some(
                    1000,
                ),
                uname: None,
            },
        },
    ),
]
